### ISSUE TITLE
How to filter topic words after topic modeling?

### ISSUE THREAD

**Issue by aayushee-gooru** [MaartenGr/BERTopic#1342]
Hi

I am looking for a way to filter certain words in topics so that topics are represented in a better manner after doing topic modelling.
The reason for filtering is certain topic words from the topic model contain unwanted numbers, plurals, etc. I am already using MMR as a finetuning model.
Here's an example of top 10 keywords for a few topics from my English K-12 course dataset:

['consensus', 'proposal', 'group', 'decision', 'decisions', 'meeting', 'groups', 'meetings', 'everyone', 'people']
['ing', 'ed', 'suffix', 'roots', 'suffixes', 'root', 'prefixes', 'prefix', 'affixes', 'word']
['capitalize', 'capitalization', 'names', 'capitalized', 'titles', 'proper', 'title', 'rules', 'capital', 'capitalise']
['preparing', 'teachers', 'national', '101722612882', 'policy', 'academies', 'preparation', 'reading', 'doi', '2010']
['2020', 'march', '24', '25', 'send', 'bookmark', '31', 'cancel', 'april', 'latest']
['2015', 'november', 'february', '2014', '2016', 'january', '2018', 'july', 'august', 'june']

I want to replace all the topic keywords having plurals (like groups, decisions, meetings, suffixes, prefixes, roots), different forms of the same word (like capitalize, capitalization, capitalized) and having numbers (like 101722612882, 2010, 2020, 24, 25, 31) above with keywords having the next best probability for that topic. Or is it better to do lemmatization and remove numbers from the dataset as a part of preprocessing. Or if there is a way to supply the model with a dictionary to have only these keywords as part of topics. 

Also, I did get a topic that did not have enough keywords for its representation:
['memory', 'sweet', 'quoting', 'eg', 'search', 'phrases', '', '', '', '']
I was not able to understand why this could have happened.

Please let me know any possible ways to overcome these issues.

Thanks!

**Comment by MaartenGr** [comment]
I think [`KeyBERTInspired`](https://maartengr.github.io/BERTopic/getting_started/representation/representation.html#keybertinspired) or [`partofspeech`](https://maartengr.github.io/BERTopic/getting_started/representation/representation.html#partofspeech) might be a better fit for you. They tend to remove unwanted numbers and plurals better than MMR typically does.

**Comment by aayushee-gooru** [comment]
Hi 
I did try both the finetuning models but output still looks the same. Is there any other way to remove unwanted numbers and plurals? Does lemmatization as preprocessing work well for BERTopic?

**Comment by MaartenGr** [comment]
Could you share your full code for trying out the PartOfSpeech? Using that method should definitely not allow for numbers.

**Comment by aayushee-gooru** [comment]
I re-ran the code and analyzed the outputs again with POS. I don't see any numbers in the topic words just after I run topic model but after removing outliers and updating the topics, the topic words are different which include numbers as well.
Seems like the effect of representation model goes away after I reduce the number of outliers and update the topics.

Could you please suggest how do I handle this? Also, the plurals and empty topic keyword issues still persist with POS representation model.

Here's the code:

`
# Topic model
from bertopic import BERTopic
# Dimension reduction
from umap import UMAP
from bertopic.representation import MaximalMarginalRelevance, KeyBERTInspired, PartOfSpeech

# Initiate UMAP
umap_model = UMAP(n_neighbors=15,
                  n_components=5,
                  min_dist=0.0,
                  metric='cosine',
                  random_state=100)

# Create representation model
pos_model = PartOfSpeech("en_core_web_sm")
representation_models = pos_model

# Initiate BERTopic
topic_model = BERTopic(verbose=True, umap_model=umap_model, language="english", calculate_probabilities=True, representation_model = representation_models)

# Run BERTopic model
topics, probs = topic_model.fit_transform(chunks_df['text_without_stopwords'])

#Reduce outliers
new_topics_1 = topic_model.reduce_outliers(chunks_df['text_without_stopwords'], topics, probabilities=probs, strategy="probabilities")

#Update topics after outlier reduction
topic_model.update_topics(chunks_df['text_without_stopwords'], topics=new_topics_1)

`

**Comment by MaartenGr** [comment]
When you use `.update_topics` you will reset the topic representation unless you specifically define the `representation_model` that you defined before. In other words, run this:

```python
topic_model.update_topics(chunks_df['text_without_stopwords'], topics=new_topics_1, representation_model=representation_models)
```

You can find more about that through the [docstrings](https://maartengr.github.io/BERTopic/api/bertopic.html#bertopic._bertopic.BERTopic.update_topics) and the [documentation](https://maartengr.github.io/BERTopic/getting_started/topicrepresentation/topicrepresentation.html)

**Comment by aayushee-gooru** [comment]
Thanks for sharing this. I realized my topic representations were not getting updated after removing outliers because I did not define the representation model in update_topics(). This works well for KeyBERTInspired and MMR Model but when i do the same for POS model, I kept getting an index error which I resolved by adding outliers as 0 in the topic model. [topic_model._outliers = 0]

But I still see few topics that have words that are blank:

['login', 'member', 'administrator', 'credentials', 'navigation', 'use', '', '', '', '']
['incovenience', 'time', '', '', '', '', '', '', '', '']
['icon', 'option', 'worksheet', 'computer', 'link', 'select', 'next', '', '', '']
['memory', 'search', 'phrases', '', '', '', '', '', '', '']
['login', 'member', 'administrator', 'credentials', 'navigation', 'use', '', '', '', '']

There are a few topics that look quite similar also [have almost identical topic words] - I intend to handle this by increasing the min_topic_size or setting nr_topics = auto.

It would be great if you could share what could be the reason for blank topic words and if there is a way to handle this issue.
Thanks!

**Comment by MaartenGr** [comment]
That might happen when there are not a sufficient number of documents in a topic or when the POS tags are set too strict. What happens here is that the top n words are checked for their POS tag but if just a few match the tag then only a few will be added. You can increase the change of getting words by increasing the `top_n_words` before passing the POS tagger.

**Comment by MaartenGr** [comment]
Closing this due to inactivity. Let me know if you want to re-open the issue!
